### DISCUSSION TITLE
How to tune gif encoding to achieve similar level of optimization of libvips?

### DISCUSSION THREAD

**Discussion by lanyusan** [SixLabors/ImageSharp#2292] [Q&A]
I have just done a simple benchmarking image frames concatenation to generate gif and gif output encoding between imagesharp against libvips 

Libvips did very well in terms of output size and colors, while imagesharp did very well in speed.

Please find this thread in libvips repo. 

https://github.com/libvips/libvips/discussions/3146

I understand imagesharp allows fine tuning of gif encoding and color sampling but I couldn't figure out how from documentation.

Is there some sample code that could lead to optimized output of size and color quality similar to libvips?

**Comment by JimBobSquarePants** [comment]
Do you have the input images and sample code? I'd like to experiment with the changes I'm working on in #2289 

There's many options all related to the Quantizer and sampling strategy passed to the encoder. 

https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.Formats.Gif.GifEncoder.html#SixLabors_ImageSharp_Formats_Gif_GifEncoder_Quantizer

**Comment by lanyusan** [comment]
The code is part of a proprietary product under development. I am sorry I cannot share it now. 

I can provide some sample images for your testing. 

And the algorithm i used to generate the gifs are also very simple:

I used three gifs and a couple of background pngs. The program randomly pick a background png, use it as a one frame gif, and then load all frames of the two gifs, then it uses the gif with most frames as base, repeat the png frame to fill a frame arrary of the base gif's frame size, and also repeat the other gif's frames to fill another array same size of the base gif's frames. 

Then overlay each frame of the three gifs' frames to make frames for the final gif. 

Then I used the default gif encoder of imagesharp and also libvips to generate output gifs.

Unfortunately the background pngs were generated randomly. I cannot get the exact same ones. But I found two similar to the ones I used in the benchmark.

Here are the images for your testing.

![100](https://user-images.githubusercontent.com/56706512/200537913-04343f15-375a-469c-8ac0-52c0dbc590ce.png)
![73](https://user-images.githubusercontent.com/56706512/200538039-795c3946-0a86-44a2-821e-b6f24d7ca3aa.png)
![1](https://user-images.githubusercontent.com/56706512/200538204-965ca091-e5b2-4ba1-b61e-be909fb948ee.gif)
![2](https://user-images.githubusercontent.com/56706512/200538284-e092f48a-27bb-4e96-bc5c-b0cb4fcf11b1.gif)
![1](https://user-images.githubusercontent.com/56706512/200538301-ba6e2115-34a0-47a3-808f-540f9edefbc0.gif)

**Reply by JimBobSquarePants** [reply]
The new decoder changes in the open PR should help but I can give you a tip right now that should vastly improve both the size and the output quality.

You should only use the background png as the root frame. There's no need to repeat it. That will remove most of the dithering noise. 

If the [`GifDisposalMethod`](https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.Formats.Gif.GifDisposalMethod.html) is set to `GifDisposalMethod.NotDispose` for the first frame containing the background and all following frames are set to `GifDisposalMethod.RestoreToPrevious` then you can achieve the same effect while encoding far less data.

The example on the right on this page shows what I mean.

http://www.theimage.com/animation/pages/disposal2.html

I'm assuming you already know how to access the gif frame metadata?

**Reply by lanyusan** [reply]
> The new decoder changes in the open PR should help but I can give you a tip right now that should vastly improve both the size and the output quality.
> 
> You should only use the background png as the root frame. There's no need to repeat it. That will remove most of the dithering noise.
> 
> If the [`GifDisposalMethod`](https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.Formats.Gif.GifDisposalMethod.html) is set to `GifDisposalMethod.NotDispose` for the first frame containing the background and all following frames are set to `GifDisposalMethod.RestoreToPrevious` then you can achieve the same effect while encoding far less data.
> 
> The example on the right on this page shows what I mean.
> 
> http://www.theimage.com/animation/pages/disposal2.html
> 
> I'm assuming you already know how to access the gif frame metadata?

Thanks for the suggestion. Will try it later.
